snapcraft/commands/remote.py

```diff
@@ -0,0 +1,293 @@
+        )
+
+        if not parsed_args.launchpad_accept_public_upload and not confirm_with_user(
+            _CONFIRMATION_PROMPT, default=False
+        ):
+            emit.progress(
+                "Remote build needs explicit acknowledgement that data sent to build servers is "
+                "public.\n"
+                "In non-interactive runs, please use the option "
+                "`--launchpad-accept-public-upload`.",
+                permanent=True,
+            )
+            return 77
+
+        builder = self._services.remote_build
+        project = cast(models.Project, self._services.project)
+        config = cast(dict[str, Any], self.config)
+        project_dir = (
+            Path(config.get("global_args", {}).get("project_dir") or ".")
+            .expanduser()
+            .resolve()
+        )
+
+        emit.trace(f"Project directory: {project_dir}")
+
+        build_planner = self._app.BuildPlannerClass.unmarshal(project.marshal())
+        full_build_plan = build_planner.get_build_plan()
+        possible_build_plan = filter_plan(
+            full_build_plan,
+            platform=parsed_args.platform,
+            build_for=parsed_args.build_for,
+            host_arch=None,
+        )
+
+        architectures: list[str] | None = list(
+            {build_info.build_for for build_info in possible_build_plan}
+        )
+
+        if parsed_args.platform and not architectures:
+            emit.progress(
+                f"platform '{parsed_args.platform}' is not present in the build plan.",
+                permanent=True,
+            )
+            return 78  # Configuration error
+
+        if parsed_args.build_for and not architectures:
+            if parsed_args.build_for in SUPPORTED_ARCHS:
+                # allow the user to build for a single architecture
+                # if the snapcraft.yaml not defined the platforms
+                architectures = [parsed_args.build_for]
+            else:
+                emit.progress(
+                    f"build-for '{parsed_args.build_for}' is not supported.",
+                    permanent=True,
+                )
+                return 78  # Configuration error
+
+        emit.debug(f"Architectures to build: {architectures}")
+
+        if not architectures:
+            architectures = None
+
+        if parsed_args.launchpad_timeout:
+            emit.debug(f"Setting timeout to {parsed_args.launchpad_timeout} seconds")
+            builder.set_timeout(parsed_args.launchpad_timeout)
+
+        build_id = get_build_id(self._app.name, project.name, project_dir)
+        if parsed_args.recover:
+            emit.progress(f"Recovering build {build_id}")
+            builds = builder.resume_builds(build_id)
+        else:
+            emit.progress(
+                "Starting new build. It may take a while to upload large projects."
+            )
+            try:
+                builds = builder.start_builds(project_dir, architectures=architectures)
```

Comment:
Other than the Launchpad side (Launchpad doesn't currently support limiting the architectures of a charm build), can you think of any reason not to include this in Charmcraft?

(Overall goal: move this class to `craft-application` and provide application-specific override hooks)

Reply:
No, I do not see any reason other than the parameters are different.